src/main/java/com/google/devtools/build/lib/cmdline/BazelStarlarkContext.java

```diff
@@ -94,19 +96,31 @@ public void storeInThread(StarlarkThread thread) {
 
   // TODO(b/236456122): Eliminate Phase, migrate analysisRuleLabel to a separate context class.
   private final Phase phase;
+  @Nullable private final InterruptibleSupplier<RepositoryMapping> mainRepoMappingSupplier;
 
   /**
    * @param phase the phase to which this Starlark thread belongs
    */
-  public BazelStarlarkContext(Phase phase) {
+  public BazelStarlarkContext(
+      Phase phase, @Nullable InterruptibleSupplier<RepositoryMapping> mainRepoMappingSupplier) {
```

Comment:
I chatted to @brandjon about this; we kind of agreed on an eventual state where all "XThreadContext" classes would inherit from this one (so that everyone has access to the main repo mapping when necessary), and this is a good location for this class (in cmdline). We'd remove all the legacy cruft from this class (all the stale comments, all `Phase`-related stuff). So overall LGTM.

One thing I don't quite understand -- why do we need a supplier here? I couldn't find a place where this wasn't just `() -> somethingReadilyAvailable`.

Reply:
A supplier is passed here:
https://github.com/bazelbuild/bazel/pull/21963/files#diff-ac21f89877bbf41328217091e31e712b170c41262802b0f4b4d8050dd9ce0e1cR36

Last time I added an unconditional dependency on the main repo mapping to every rule, my commit was rolled back, so I won't try that again 🙂

Reply:
oh, I see. So the rule would only depend on the main repo mapping if its impl function prints a label? I suppose that's fine enough.

Reply:
Yes, exactly. That should reduce the memory impact dramatically.